ds701_book/26-TimeSeries.qmd

```diff
@@ -0,0 +1,1351 @@
+seasonal_ts = detrended_ts.groupby(detrended_ts.index.month).mean()
+seasonal_ts.plot(figsize=(8, 3), label='Seasonal')
+plt.ylabel('Number of Passengers')
+plt.title('Air Passengers 1949-1960')
+plt.legend()
+plt.show()
+```
+
+---
+
+We can also look at the box plot of the detrended series for each month.
+
+```{python}
+#| fig-align: center
+# Box plot of the detrended series for each month
+import seaborn as sns
+
+# Create a DataFrame with the detrended series and the corresponding month
+detrended_df = pd.DataFrame({'Detrended': detrended_ts, 'Month': detrended_ts.index.month})
+
+# Plot the box plot
+plt.figure(figsize=(8, 4))
+sns.boxplot(x='Month', y='Detrended', data=detrended_df, palette='tab10', hue='Month', legend=False)
+plt.xlabel('Month')
+plt.ylabel('Detrended Number of Passengers')
+plt.title('Box Plot of Detrended Air Passengers by Month')
+plt.show()
+```
+
+Clearly, certain months have more variability than others.
+
+## Remove the seasonal component
+
+Now we can subtract the seasonal component to get the irregular component.
+
+```{python}
+#| code-fold: false
+# Broadcast the seasonal component to the length of the detrended series
+# and subtract it from the detrended series
+
+irregular_ts = detrended_ts - np.tile(seasonal_ts, len(detrended_ts) // len(seasonal_ts))
+```
+
+And plot the residual irregular component.
+
+```{python}
+#| fig-align: center
+plt.figure(figsize=(8, 3))
+irregular_ts.plot(label='Irregular', color='red')
+plt.ylabel('Irregular')
+plt.title('Irregular Component')
+plt.legend()
+plt.show()
+```
+
+**Question**: What do you observe about the irregular component? Is it truly
+random?
+
+---
+
+Finally, we can put the detrended, seasonal, and irregular plots together.
+
+```{python}
+#| fig-align: center
+# Plot detrended, seasonal, and irregular components
+plt.figure(figsize=(8, 6))
+
+# Plot detrended series
+plt.subplot(3, 1, 1)
+detrended_ts.plot(label='Detrended', color='blue')
+plt.ylabel('Detrended')
+plt.title('Detrended Component')
+plt.legend()
+
+# Plot seasonal series
+plt.subplot(3, 1, 2)
+seasonal_ts.plot(label='Seasonal', color='green')
+plt.ylabel('Seasonal')
+plt.title('Seasonal Component')
+plt.legend()
+
+# Plot irregular series
+plt.subplot(3, 1, 3)
+irregular_ts.plot(label='Irregular', color='red')
+plt.ylabel('Irregular')
+plt.title('Irregular Component')
+plt.legend()
+
+plt.tight_layout()
+plt.show()
+
+```
+
+
+## Seasonal and Trend decomposition using Loess (STL)
+
+- **STL** is a more flexible and robust method for decomposing time series data.
+- It uses locally weighted regression (Loess) to estimate the trend and seasonal components [@cleveland1990stl].
+- **Advantages**:
+  - Handles any type of seasonality (e.g., weekly, monthly).
+  - Can deal with missing values and outliers.
+  - Provides smooth and adaptable trend and seasonal components.
+
+> See [Course Notes](#locally-estimated-scatterplot-smoothing----loessgpt4o) for a more detailed description of Loess.
```

Comment:
This links only to a local file on my machine and not to the course webpage.

Reply:
It is a relative link which should work when the website is published, but I can't test it till then.

Reply:
Ok. No problem. Just was confused since it didn't work for me just yet. Thanks!